src/installer/pkg/sfx/Microsoft.NETCore.App/Microsoft.NETCore.App.Crossgen2.sfxproj

```diff
@@ -13,56 +14,66 @@
-    <TargetSpec>$(TargetOSComponent)-$(TargetArchitecture)</TargetSpec>
-  </PropertyGroup>
+  <Target Name="PublishCrossgen"
+          BeforeTargets="GetFilesToPackage">
+
+    <!-- Copy System.Private.CoreLib from the coreclr bin directory to the runtime pack directory,
+         as we always need the copy of System.Private.CoreLib that matches exactly with the runtime. -->
+    <Copy SourceFiles="$(CoreCLRArtifactsPath)System.Private.CoreLib.dll"
+          DestinationFolder="$(MicrosoftNetCoreAppRuntimePackNativeDir)"
+          SkipUnchangedFiles="true" />
 
-  <ItemGroup>
-    <NativeRuntimeAsset Include="$(CoreCLRCrossgen2Dir)crossgen2$(ExeSuffix)" TargetPath="tools/" />
-    <Reference Include="$(CoreCLRCrossgen2Dir)crossgen2.dll" />
-    <Reference Include="$(CoreCLRCrossgen2Dir)ILCompiler*.dll" />
-    <Reference Condition="'$(DotNetBuildFromSource)' != 'true'" Include="$(CoreCLRCrossgen2Dir)Microsoft.DiaSymReader.dll" />
-    <NativeRuntimeAsset Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)jitinterface_$(TargetArchitecture)$(LibSuffix)" TargetPath="tools/" />
-    <NativeRuntimeAsset Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)clrjit_win_x86_$(TargetArchitecture)$(LibSuffix)" TargetPath="tools/"  />
-    <NativeRuntimeAsset Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)clrjit_universal_arm_$(TargetArchitecture)$(LibSuffix)" TargetPath="tools/"  />
-    <NativeRuntimeAsset Condition="'$(TargetArchitecture)' == 'arm64' or  '$(TargetArchitecture)' == 'x64'" Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)clrjit_win_x64_$(TargetArchitecture)$(LibSuffix)"  TargetPath="tools/" />
-    <NativeRuntimeAsset Condition="'$(TargetArchitecture)' == 'arm64' or  '$(TargetArchitecture)' == 'x64'" Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)clrjit_unix_x64_$(TargetArchitecture)$(LibSuffix)" TargetPath="tools/" />
-    <NativeRuntimeAsset Condition="'$(TargetArchitecture)' == 'arm64' or  '$(TargetArchitecture)' == 'x64'" Include="$(CoreCLRCrossgen2Dir)$(LibPrefix)clrjit_universal_arm64_$(TargetArchitecture)$(LibSuffix)" TargetPath="tools/"  />
-    <!-- Include the native hosting layer  -->
-    <NativeRuntimeAsset Include="$(DotNetHostBinDir)/$(LibPrefix)hostfxr$(LibSuffix)" TargetPath="tools/" />
-    <NativeRuntimeAsset Include="$(DotNetHostBinDir)/$(LibPrefix)hostpolicy$(LibSuffix)" TargetPath="tools/" />
-  </ItemGroup>
+    <PropertyGroup>
+      <_TargetsToRun>Restore;Publish;PublishItemsOutputGroup</_TargetsToRun>
+    </PropertyGroup>
+
+    <MSBuild Projects="$(RepoRoot)src/coreclr/tools/aot/crossgen2/crossgen2.csproj"
+             Properties="RunningPublish=true
+              ;RuntimeIdentifier=$(PackageRID)
+              ;CoreCLRArtifactsPath=$(CoreCLRArtifactsPath)
+              ;R2ROverridePath=$(MSBuildThisFileDirectory)ReadyToRun.targets"
+             Targets="$(_TargetsToRun)">
+      <Output TaskParameter="TargetOutputs"
+              ItemName="_RawCrossgenPublishFiles" />
+    </MSBuild>
```

Comment:
You really want to avoid invoking restore in the same task as other targets run.

The pattern that I followed over the years that worked very well is to have two MSBuild task invocations. One for restore and one for anything else. What you currently have might work but is probably fragile and might easily break in the future. I recommend to add another msbuild task before the one that you currently have:

```suggestion
<MSBuild Projects="$(RepoRoot)src/coreclr/tools/aot/crossgen2/crossgen2.csproj"
             Targets="Restore"
             Properties="MSBuildRestoreSessionId=$([System.Guid]::NewGuid())
              ;RunningPublish=true
              ;RuntimeIdentifier=$(PackageRID)
              ;CoreCLRArtifactsPath=$(CoreCLRArtifactsPath)
              ;R2ROverridePath=$(MSBuildThisFileDirectory)ReadyToRun.targets" />

    <MSBuild Projects="$(RepoRoot)src/coreclr/tools/aot/crossgen2/crossgen2.csproj"
             Targets="Publish;PublishItemsOutputGroup"
             Properties="RunningPublish=true
              ;RuntimeIdentifier=$(PackageRID)
              ;CoreCLRArtifactsPath=$(CoreCLRArtifactsPath)
              ;R2ROverridePath=$(MSBuildThisFileDirectory)ReadyToRun.targets">
      <Output TaskParameter="TargetOutputs"
              ItemName="_RawCrossgenPublishFiles" />
    </MSBuild>
```